### ISSUE TITLE
BUG: set_zlim does not set vmax as expected

### ISSUE THREAD

**Issue by vdbma** [yt-project/yt#3887]
### Bug report

**Bug summary**

The actual vmax of a plot does not match the vmax specified in `set_zlim`.

**Code for reproduction**

```python
import yt
import numpy as np

shape = (32, 16, 1)
a = np.linspace(-1, 1, 16)
b = np.ones((32, 16))
c = np.reshape(a*b, shape)
data = {("gas", "density"): c}

ds = yt.load_uniform_grid(
    data,
    shape,
    bbox=np.array([[0.0, 5.0], [0, 1], [-0.1, +0.1]]),
)
p = yt.SlicePlot(ds,"x",("gas", "density"))
p.set_zlim(("gas", "density"), -10, 10)
p.save("zmax.png")
```

**Actual outcome**

![zmax](https://user-images.githubusercontent.com/93188557/161748458-add8c3e0-b6d5-40b0-a752-ff0cd8729b5b.png)

**Version Information**

  * Operating System: Debian 5.10.106-1 x86_64 GNU/Linux
  * Python Version: 3.9.9
  * yt version: 4.0.2 (through `pip install yt`)

**Comment by neutrinoceros** [comment]
Thank you for reporting !
So, the good news is that the problem you're reporting is already fixed on the main banch (as well as on the backport branch)... the bad news is that the resulting image is now also less useful because the linthresh is autoset to zmax (or maybe -zmin)
![zmax](https://user-images.githubusercontent.com/14075922/161825931-bd5f8bf7-6123-4b43-ba22-ed0aae193b5d.png)

I note that with 5e480b72 (part of the backport branch between 4.0.2 and the current tip) we get the best of both worlds

![zmax](https://user-images.githubusercontent.com/14075922/161827077-a58a5452-ba5e-4415-b737-ada6e0da9f5e.png)

The problematic change bisects to #3859
I'll try to find a better solution that still fixes #3858 without the undesirable side effect.

**Comment by neutrinoceros** [comment]
For posterity, the original bug reported here (not having a proper tick label at user-defined zmax) was fixed in #3793
